neuralforecast/utils.py

```diff
@@ -439,10 +442,130 @@ def augment_calendar_df(df, freq="H"):
     idxs = idx.get_indexer(vals)
     missing = [v for i, v in zip(idxs, vals) if i == -1]
     if missing:
         raise ValueError(f"The following values are missing from the index: {missing}")
     return idxs
+
+# %% ../nbs/utils.ipynb 31
+class ConformalIntervals:
+    """Class for storing conformal intervals metadata information."""
+
+    def __init__(
+        self,
+        n_windows: int = 2,
+        method: str = "conformal_distribution",
+        enable_quantiles: bool = False,
+    ):
+        """
+        n_windows : int
+            Number of windows to evaluate.
+        method : str, default is conformal_distribution
+            One of the supported methods for the computation of conformal prediction:
+            conformal_error or conformal_distribution
+        enable_quantiles : bool, default is False
+            If set to True, we create prediction intervals on top of quantiled outputs, e.g. prediction made
+            with MQLoss(level=[80]) will be conformalized with the respective conformal
+            levels (prediction columns having 'model-lo/hi-80-conformal-lo/hi-#').
+        """
+        if n_windows < 2:
+            raise ValueError(
+                "You need at least two windows to compute conformal intervals"
+            )
+        allowed_methods = ["conformal_error", "conformal_distribution"]
+        if method not in allowed_methods:
+            raise ValueError(f"method must be one of {allowed_methods}")
+        self.n_windows = n_windows
+        self.method = method
+        self.enable_quantiles = enable_quantiles
+
+    def __repr__(self):
+        return f"ConformalIntervals(n_windows={self.n_windows}, method='{self.method}')"
+
+# %% ../nbs/utils.ipynb 32
+def add_conformal_distribution_intervals(
+    fcst_df: DFType,
+    cs_df: DFType,
+    model_names: List[str],
+    level: List[Union[int, float]],
+    cs_n_windows: int,
+    n_series: int,
+    horizon: int,
+) -> DFType:
+    """
+    Adds conformal intervals to a `fcst_df` based on conformal scores `cs_df`.
+    `level` should be already sorted. This strategy creates forecasts paths
+    based on errors and calculate quantiles using those paths.
+    """
+    fcst_df = ufp.copy_if_pandas(fcst_df, deep=False)
+    alphas = [100 - lv for lv in level]
+    cuts = [alpha / 200 for alpha in reversed(alphas)]
+    cuts.extend(1 - alpha / 200 for alpha in alphas)
+    for model in model_names:
+        scores = cs_df[model].to_numpy().reshape(cs_n_windows, n_series, horizon)
+        # restrict scores to horizon
+        scores = scores[:, :, :horizon]
+        mean = fcst_df[model].to_numpy().reshape(1, n_series, -1)
+        scores = np.vstack([mean - scores, mean + scores])
+        quantiles = np.quantile(
+            scores,
+            cuts,
+            axis=0,
+        )
+        quantiles = quantiles.reshape(len(cuts), -1).T
+        lo_cols = [f"{model}-conformal-lo-{lv}" for lv in reversed(level)]
```

Comment:
Naming should be without the `-conformal` tag, so that naming is identical to normal distribution outputs